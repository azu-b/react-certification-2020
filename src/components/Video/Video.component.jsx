import React from 'react';
import { Container, VideoContainer, Title, Description } from './Video.styled';

const Video = (props) => {
  const { id, title, description } = props;

  return (
    <Container>
      <VideoContainer src={`https://www.youtube.com/embed/${id}`} />
      <Title>{title}</Title>
      <Description>{description}</Description>
    </Container>
  );
};

export default Video;
