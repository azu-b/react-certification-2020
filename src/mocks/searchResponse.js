const searchResponse = {
  kind: 'youtube#searchListResponse',
  etag: 'n9Y42Gm30xxTdZ6Bc6y8YPUX454',
  nextPageToken: 'CAUQAA',
  regionCode: 'MX',
  pageInfo: {
    totalResults: 1000000,
    resultsPerPage: 5,
  },
  items: [
    {
      kind: 'youtube#searchResult',
      etag: 'Jt8LbUC1T4vFrfk8d51TBvmS9B0',
      id: {
        kind: 'youtube#video',
        videoId: '3w5hjn7sY8g',
      },
      snippet: {
        publishedAt: '2020-10-05T00:00:00Z',
        channelId: 'UCq8up0Ew9K0IiRxbz75BZsg',
        title: 'HIGH ELO Best Champions TIER List - League of Legends Patch 10.20',
        description:
          'Want instant, easily-accessible, 24-7 coaching from high elo players? Then check out our website: https://www.proguides.com/yt Patch 10.20 comes with ...',
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/3w5hjn7sY8g/default.jpg',
            width: 120,
            height: 90,
          },
          medium: {
            url: 'https://i.ytimg.com/vi/3w5hjn7sY8g/mqdefault.jpg',
            width: 320,
            height: 180,
          },
          high: {
            url: 'https://i.ytimg.com/vi/3w5hjn7sY8g/hqdefault.jpg',
            width: 480,
            height: 360,
          },
        },
        channelTitle: 'ProGuides Challenger League of Legends Guides',
        liveBroadcastContent: 'none',
        publishTime: '2020-10-05T00:00:00Z',
      },
    },
    {
      kind: 'youtube#searchResult',
      etag: 'e0RIH0Ar2o0ZTxHcaq70cavX5Xo',
      id: {
        kind: 'youtube#video',
        videoId: 'RR0JFDWuceo',
      },
      snippet: {
        publishedAt: '2020-10-04T09:00:08Z',
        channelId: 'UClh5azhOaKzdlThQFgoq-tw',
        title:
          'When LOL Players PERFECT ESCAPE... 200 IQ ESCAPES MONTAGE (League of Legends)',
        description:
          'When LOL Players PERFECT ESCAPE... Best Pro Player Escapes, Perfect Timing Moments, 200 IQ Tricks and Jukes in a League of Legends Montage ...',
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/RR0JFDWuceo/default.jpg',
            width: 120,
            height: 90,
          },
          medium: {
            url: 'https://i.ytimg.com/vi/RR0JFDWuceo/mqdefault.jpg',
            width: 320,
            height: 180,
          },
          high: {
            url: 'https://i.ytimg.com/vi/RR0JFDWuceo/hqdefault.jpg',
            width: 480,
            height: 360,
          },
        },
        channelTitle: 'Protatomonster',
        liveBroadcastContent: 'none',
        publishTime: '2020-10-04T09:00:08Z',
      },
    },
    {
      kind: 'youtube#searchResult',
      etag: '6VVytJOelo1z-FdT6KmNHjfE0AY',
      id: {
        kind: 'youtube#video',
        videoId: 'o1YfnQ5enz8',
      },
      snippet: {
        publishedAt: '2020-10-05T00:03:51Z',
        channelId: 'UCv7PqO2lW3eK16sQngYPwsA',
        title:
          'THIS ITEM MAKE SHACO PERMINENTLY INVISIBLE (NO COUNTER) - BunnyFuFuu | League Of Legends | Season 11',
        description:
          'Shaco is always invisible you gotta check this out $2500 CASH or GAMING SETUP WORLDWIDE GIVEAWAY LINK ...',
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/o1YfnQ5enz8/default.jpg',
            width: 120,
            height: 90,
          },
          medium: {
            url: 'https://i.ytimg.com/vi/o1YfnQ5enz8/mqdefault.jpg',
            width: 320,
            height: 180,
          },
          high: {
            url: 'https://i.ytimg.com/vi/o1YfnQ5enz8/hqdefault.jpg',
            width: 480,
            height: 360,
          },
        },
        channelTitle: 'BunnyFuFuu',
        liveBroadcastContent: 'none',
        publishTime: '2020-10-05T00:03:51Z',
      },
    },
    {
      kind: 'youtube#searchResult',
      etag: 'muLyNyvSkC63p1I96G_nMt62Sp4',
      id: {
        kind: 'youtube#video',
        videoId: 'aR-KAldshAE',
      },
      snippet: {
        publishedAt: '2020-01-09T12:00:16Z',
        channelId: 'UC2t5bjwHdUX4vM2g8TRDq5g',
        title:
          'Warriors | Season 2020 Cinematic - League of Legends (ft. 2WEI and Edda Hayes)',
        description:
          'Warriors: We are warriors. Season 2020 has begun. LISTEN NOW: https://smarturl.it/warriorslol Spotify: http://smarturl.it/warriorslol/spotify Apple Music: ...',
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/aR-KAldshAE/default.jpg',
            width: 120,
            height: 90,
          },
          medium: {
            url: 'https://i.ytimg.com/vi/aR-KAldshAE/mqdefault.jpg',
            width: 320,
            height: 180,
          },
          high: {
            url: 'https://i.ytimg.com/vi/aR-KAldshAE/hqdefault.jpg',
            width: 480,
            height: 360,
          },
        },
        channelTitle: 'League of Legends',
        liveBroadcastContent: 'none',
        publishTime: '2020-01-09T12:00:16Z',
      },
    },
    {
      kind: 'youtube#searchResult',
      etag: 'dJDFTqE2xU-_WNKHcVBC4NkbT4g',
      id: {
        kind: 'youtube#video',
        videoId: 'pngf7xqrsDU',
      },
      snippet: {
        publishedAt: '2020-10-04T12:00:10Z',
        channelId: 'UCq8up0Ew9K0IiRxbz75BZsg',
        title:
          'NEW UPDATES: Seraphine Abilities LEAKED? + Upcoming JG &amp; Mid Nerfs &amp; MORE - League of Legends',
        description:
          'Want instant, easily-accessible, 24-7 coaching from high elo players? Then check out our website: https://www.proguides.com/yt Come join our Discord server ...',
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/pngf7xqrsDU/default.jpg',
            width: 120,
            height: 90,
          },
          medium: {
            url: 'https://i.ytimg.com/vi/pngf7xqrsDU/mqdefault.jpg',
            width: 320,
            height: 180,
          },
          high: {
            url: 'https://i.ytimg.com/vi/pngf7xqrsDU/hqdefault.jpg',
            width: 480,
            height: 360,
          },
        },
        channelTitle: 'ProGuides Challenger League of Legends Guides',
        liveBroadcastContent: 'none',
        publishTime: '2020-10-04T12:00:10Z',
      },
    },
  ],
};

export default searchResponse;
