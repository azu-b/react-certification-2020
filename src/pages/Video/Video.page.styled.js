import styled from 'styled-components';

const Container = styled.div`
  margin-top: 60px;
  display: flex;
  flex-direction: column;

  @media (min-width: 1024px) {
    flex-direction: row;
  }
`;

const VideoContainer = styled.div`
  width: 100%;
  padding: 10px;

  @media (min-width: 1024px) {
    width: 60%;
    padding: 20px;
  }
`;

const RelatedVideosContainer = styled.div`
  width: 100%;
  padding: 10px;

  @media (min-width: 1024px) {
    width: 40%;
    padding: 20px;
  }
`;

const Error = styled.div`
  font-family: 'Folks Normal';
  color: #9c0000;
  font-size: 32px;
  margin: 20px;

  @media (min-width: 768px) {
    font-size: 36px;
    margin: 20px auto;
  }
`;

const Load = styled.div`
  font-family: 'Heebo Regular';
  color: #0093ff;
  font-size: 28px;
  margin: 20px;

  @media (min-width: 768px) {
    font-size: 32px;
    margin: 20px auto;
  }
`;

const LoadRelated = styled(Load)`
  font-size: 20px;
  margin: 10px;

  @media (min-width: 768px) {
    font-size: 24px;
    margin: 10px auto;
  }
`;

export { Container, VideoContainer, RelatedVideosContainer, Error, Load, LoadRelated };
