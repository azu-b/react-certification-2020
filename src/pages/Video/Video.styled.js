import styled from 'styled-components';

const Container = styled.div`
  margin-top: 60px;
  display: flex;
  justify-content: center;
  align-items: flex-start;
`;

export { Container };
