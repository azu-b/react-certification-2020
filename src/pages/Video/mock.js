const mockVideo = {
  id: 'uE-1RPDqJAY',
  description: 'THEY"RE TAKING THE HOBBITS TO ISENGARD (not my work)',
  title: 'taking the hobbits to isengard',
};

export default mockVideo;
