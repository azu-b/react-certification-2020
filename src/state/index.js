export { default as SearchContext } from './Search.context';
export { default as useAuth } from './Auth.context';
export { default as AuthProvider } from './Auth.provider';
