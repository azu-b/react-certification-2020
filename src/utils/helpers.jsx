import React from 'react';
import { Router, Route } from 'react-router-dom';
import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';

const renderWithRouterMatch = (
  ui,
  {
    path = '/',
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {}
) => {
  return {
    ...render(
      <Router history={history}>
        <Route path={path} component={ui} />
      </Router>
    ),
  };
};

const cleanYouTubeResponse = (videos) => {
  const cleanedVideos = videos.map((video) => {
    return {
      id: video?.id?.videoId || '',
      title: video?.snippet?.title || '',
      description: video?.snippet?.description || '',
      thumbnails: video?.snippet?.thumbnails || '',
    };
  });

  return cleanedVideos;
};

const storage = {
  get(key) {
    try {
      const rawValue = window.localStorage.getItem(key);
      return JSON.parse(rawValue);
    } catch (error) {
      console.error(`Error parsing storage item "${key}".`);
      return null;
    }
  },

  set(key, value) {
    window.localStorage.setItem(key, JSON.stringify(value));
  },
};

export { renderWithRouterMatch, cleanYouTubeResponse, storage };
