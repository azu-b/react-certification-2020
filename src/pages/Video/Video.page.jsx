import React, { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import VideoComponent from '../../components/Video';
import VideoList from '../../components/VideoList';
import { Container, VideoContainer, RelatedVideosContainer } from './Video.styled';
import { YouTubeAPI, cleanYouTubeResponse } from '../../utils/youtube';

const Video = () => {
  const { id: videoId } = useParams();
  const [video, setVideo] = useState(undefined);
  const [relatedVideos, setRelatedVideos] = useState(undefined);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const getVideoInfo = async () => {
      try {
        const response = await YouTubeAPI.get('/videos', {
          params: {
            id: videoId,
          },
        });
        const cleanVideos = cleanYouTubeResponse(response.data.items);
        console.log(cleanVideos);
        setVideo(cleanVideos[0]);
        setIsLoading(false);
      } catch (error) {
        console.log(error);
      }
    };

    const getRelatedVideos = async () => {
      try {
        const response = await YouTubeAPI.get('/search', {
          params: {
            relatedToVideoId: videoId,
            type: 'video',
            maxResults: 25,
          },
        });
        const cleanVideos = cleanYouTubeResponse(response.data.items);
        setRelatedVideos(cleanVideos);
      } catch (error) {
        console.log(error);
      }
    };

    getVideoInfo();
    getRelatedVideos();
  }, [videoId]);

  return (
    <Container>
      {isLoading && <div>Loading video info...</div>}
      {video ? (
        <VideoContainer>
          <VideoComponent
            id={videoId}
            title={video.title}
            description={video.description}
          />
        </VideoContainer>
      ) : (
        <div>Ups! No video with this id found</div>
      )}
      {relatedVideos && (
        <RelatedVideosContainer>
          <VideoList videos={relatedVideos} />
        </RelatedVideosContainer>
      )}
    </Container>
  );
};

export default Video;
