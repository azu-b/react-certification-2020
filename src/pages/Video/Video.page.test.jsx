import React from 'react';
import { cleanup, act, screen } from '@testing-library/react';
import pretty from 'pretty';
import { AuthProvider } from '../../state';
import Video from './index';
import YouTubeAPI from '../../utils/youtube';
import { renderWithRouterMatch } from '../../utils/helpers';
import searchResponse from '../../mocks/searchResponse';

jest.mock('../../utils/youtube');

describe('<VideoPage>', () => {
  const VideoWithAuthProvider = () => (
    <AuthProvider>
      <Video />
    </AuthProvider>
  );

  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  it('renders', async () => {
    YouTubeAPI.get.mockImplementation(() => Promise.resolve({ data: searchResponse }));

    await act(async () => {
      renderWithRouterMatch(VideoWithAuthProvider, {
        route: '/123',
        path: '/:id',
      });
    });

    const video = screen.getByTestId('video');
    const description = screen.getByTestId('description');
    const titles = screen.getAllByTestId('title');
    const links = screen.getAllByTestId('link');
    const thumbnails = screen.getAllByTestId('thumbnail');
    const relatedVideosTitle = screen.getByText('Related Videos');

    expect(video.tagName).toBe('IFRAME');
    expect(description.tagName).toBe('P');
    expect(titles).toHaveLength(searchResponse.items.length + 1);
    expect(links).toHaveLength(searchResponse.items.length);
    expect(thumbnails).toHaveLength(searchResponse.items.length);
    expect(titles[0].tagName).toBe('H1');
    expect(titles[1].tagName).toBe('H4');
    expect(links[0].tagName).toBe('A');
    expect(thumbnails[0].tagName).toBe('IMG');
    expect(relatedVideosTitle.tagName).toBe('H2');
  });

  it('matches snapshot', async () => {
    YouTubeAPI.get.mockImplementation(() => Promise.resolve({ data: searchResponse }));

    let wrapper;

    await act(async () => {
      wrapper = renderWithRouterMatch(VideoWithAuthProvider, {
        route: '/123',
        path: '/:id',
      });
    });

    expect(pretty(wrapper.container.innerHTML)).toMatchSnapshot();
  });

  it('renders an error message when video id is invalid', async () => {
    YouTubeAPI.get.mockImplementationOnce(() =>
      Promise.reject(new Error('Video ID is invalid'))
    );

    let wrapper;

    await act(async () => {
      wrapper = renderWithRouterMatch(VideoWithAuthProvider, {
        route: '/invalid',
        path: '/:id',
      });
    });

    expect(wrapper.container.innerHTML).toMatch('Ups! No video with this id was found');
  });

  it('renders an error message when there are no related videos found', async () => {
    YouTubeAPI.get
      .mockImplementationOnce(() => Promise.resolve({ data: searchResponse }))
      .mockImplementationOnce(() => Promise.reject(new Error('No related videos found')));

    let wrapper;

    await act(async () => {
      wrapper = renderWithRouterMatch(VideoWithAuthProvider, {
        route: '/123',
        path: '/:id',
      });
    });

    expect(wrapper.container.innerHTML).toMatch('No related videos were found');
  });
});
