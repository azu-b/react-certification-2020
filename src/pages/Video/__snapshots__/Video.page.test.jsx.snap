// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoPage> matches snapshot 1`] = `
"<div class=\\"sc-crzoUp bluRdc\\">
  <div class=\\"sc-dIsAE dmHtcx\\">
    <div class=\\"sc-bdnylx jiVbiK\\"><iframe data-testid=\\"video\\" src=\\"https://www.youtube.com/embed/123\\" class=\\"sc-gtssRu llVzmO\\"></iframe>
      <h1 data-testid=\\"title\\" class=\\"sc-dlnjPT cuonUe\\">HIGH ELO Best Champions TIER List - League of Legends Patch 10.20</h1>
      <p data-testid=\\"description\\" class=\\"sc-hKFyIo ckbdfc\\">Want instant, easily-accessible, 24-7 coaching from high elo players? Then check out our website: https://www.proguides.com/yt Patch 10.20 comes with ...</p>
    </div>
  </div>
  <div class=\\"sc-bqGHjH kZvmdo\\">
    <div class=\\"sc-iCoHVE jowieF\\">
      <h2 class=\\"sc-gKAblj epZClN\\">Related Videos</h2>
      <div data-testid=\\"video-list\\" class=\\"sc-iCoHVE sc-fujyUd jowieF dJiQVZ\\"><a data-testid=\\"link\\" href=\\"/video/3w5hjn7sY8g\\">
          <div data-testid=\\"video-item\\" class=\\"sc-pNWxx kUTBDY\\"><img data-testid=\\"thumbnail\\" src=\\"https://i.ytimg.com/vi/3w5hjn7sY8g/default.jpg\\" class=\\"sc-jrsJCI fFewxX\\">
            <h4 data-testid=\\"title\\" class=\\"sc-kEqYlL cKACju\\">HIGH ELO Best Champions TIER List - League of Legends Patch 10.20</h4>
          </div>
        </a><a data-testid=\\"link\\" href=\\"/video/RR0JFDWuceo\\">
          <div data-testid=\\"video-item\\" class=\\"sc-pNWxx kUTBDY\\"><img data-testid=\\"thumbnail\\" src=\\"https://i.ytimg.com/vi/RR0JFDWuceo/default.jpg\\" class=\\"sc-jrsJCI fFewxX\\">
            <h4 data-testid=\\"title\\" class=\\"sc-kEqYlL cKACju\\">When LOL Players PERFECT ESCAPE... 200 IQ ESCAPES MONTAGE (League of Legends)</h4>
          </div>
        </a><a data-testid=\\"link\\" href=\\"/video/o1YfnQ5enz8\\">
          <div data-testid=\\"video-item\\" class=\\"sc-pNWxx kUTBDY\\"><img data-testid=\\"thumbnail\\" src=\\"https://i.ytimg.com/vi/o1YfnQ5enz8/default.jpg\\" class=\\"sc-jrsJCI fFewxX\\">
            <h4 data-testid=\\"title\\" class=\\"sc-kEqYlL cKACju\\">THIS ITEM MAKE SHACO PERMINENTLY INVISIBLE (NO COUNTER) - BunnyFuFuu | League Of Legends | Season 11</h4>
          </div>
        </a><a data-testid=\\"link\\" href=\\"/video/aR-KAldshAE\\">
          <div data-testid=\\"video-item\\" class=\\"sc-pNWxx kUTBDY\\"><img data-testid=\\"thumbnail\\" src=\\"https://i.ytimg.com/vi/aR-KAldshAE/default.jpg\\" class=\\"sc-jrsJCI fFewxX\\">
            <h4 data-testid=\\"title\\" class=\\"sc-kEqYlL cKACju\\">Warriors | Season 2020 Cinematic - League of Legends (ft. 2WEI and Edda Hayes)</h4>
          </div>
        </a><a data-testid=\\"link\\" href=\\"/video/pngf7xqrsDU\\">
          <div data-testid=\\"video-item\\" class=\\"sc-pNWxx kUTBDY\\"><img data-testid=\\"thumbnail\\" src=\\"https://i.ytimg.com/vi/pngf7xqrsDU/default.jpg\\" class=\\"sc-jrsJCI fFewxX\\">
            <h4 data-testid=\\"title\\" class=\\"sc-kEqYlL cKACju\\">NEW UPDATES: Seraphine Abilities LEAKED? + Upcoming JG &amp;amp; Mid Nerfs &amp;amp; MORE - League of Legends</h4>
          </div>
        </a></div>
    </div>
  </div>
</div>"
`;
